site/content/blog/2020-03-16-advanced-git-tbilisi-free-university.md

```diff
@@ -0,0 +1,437 @@
+git commit -m 'wip'
+go test ./... -coverprofile=coverage.out && go tool cover -html=coverage.out -o coverage.html && open coverage.html
+```
+
+Excellent! Now let's look at the commit history:
+
+```sh
+git log -p
+```
+
+```sh
+commit 3489625af31bdf825f615e1963220c0857d7ca31 (HEAD -> test-git-rebase)
+Author: Jesse Hallam 
+Date:   Mon Mar 16 11:31:05 2020 -0300
+
+    wip
+
+commit f1ef494d443099ebbbb719286ddef2eb8a80168e
+Author: Jesse Hallam 
+Date:   Mon Mar 16 11:26:09 2020 -0300
+
+    remove dead code
+
+commit b57fde46986f27197696e72e93493a98896baaea
+Author: Jesse Hallam 
+Date:   Mon Mar 16 11:27:37 2020 -0300
+
+    improve testing coe
+
+commit da0d04f00b9b11b10969a3467f88d109abc0ab9b
+Author: Jesse Hallam 
+Date:   Mon Mar 16 11:24:53 2020 -0300
+
+    add TestConfigEnableDeveloper
+```
+
+So the final diff is great, but the commits themselves are pretty disorganized. That's not how I want to present my work to my team. Imagine if these were dozens of commits over hundreds of changes. Imagine if I decided to rename a variable used all over the code half-way through my changes. It's pretty hard to review a set of disorganized changes. `git rebase` to the rescue! Let's run it interactively:
+
+```sh
+git rebase -i origin/master
+```
+
+It's thrown me into an editor, with each of my commits as a separate line. By default, they are all (cherry) "picked" to be kept, but I have a number of options to help me reorganize my code. So the first two commits and the last commit really don't need to be separated at all: they should just be a single commit. I'm going to move the last line up above the third commit, and then tell `git rebase` to squash the two commits into the first commit:
+
+```sh
+pick … (Jesse Hallam) add TestConfigEnableDeveloper
+squash … (Jesse Hallam) improve testing coe
+squash … (Jesse Hallam) wip
+pick  … (Jesse Hallam) remove dead code
+```
+
+When I save and exit this file, git will create a temporary ref at origin/master, and cherry-pick the changes in the order I requested. When I tell it to squash, commit amends to the previous commit, effectively squishing the three commits together. Save and exit, and you'll see I'm being prompted to write a new commit history for the squash commits. Finally, let's look at the diff:
+
+```sh
+git log -p 
+```
+
+Much better! Except, the messages are still pretty disorganized. Let's fix that with another rebase, but we'll reword the messages this time:
+
+```sh
+git rebase -i origin/master
+```
+
+```sh
+reword … (Jesse Hallam) add TestConfigEnableDeveloper
+pick  … (Jesse Hallam) remove dead code
+```
+
+Neat! Of course, I didn't have to do this in two passes, but I wanted to show you that you can run it as often as necessary. Sometimes, when you reorder your commits, you'll end up with merge conflicts. Just as with `git cherry-pick` and `git revert`, `git rebase` will stop and ask you to resolve the conflicts. Then you can --continue and it will pick up with the next commit.
```

Comment:
```suggestion
Neat! Of course, I didn't have to do this in two passes, but I wanted to show you that you can run it as often as necessary. Sometimes, when you reorder your commits, you'll end up with merge conflicts. Just as with `git cherry-pick` and `git revert`, `git rebase` will stop and ask you to resolve the conflicts. Then you can `--continue` and it will pick up with the next commit.
```